src/browser/pages/vscode.ts

```diff
@@ -54,6 +54,84 @@ export function getNlsConfiguration(document: Document) {
   return JSON.parse(nlsConfig) as NlsConfiguration
 }
 
+type RegisterRequireOnSelfType = {
+  // NOTE@jsjoeio
+  // We get the self type by looking at window.self.
+  self: Window & typeof globalThis
+  window: Window
+  nlsConfig: NlsConfiguration
+  options: Options
+}
+
+type RequireOnSelfType = {
+  baseUrl: string
+  recordStats: boolean
+  paths: {
+    [key: string]: string
+  }
+  "vs/nls": NlsConfiguration
+}
+
+/**
+ * A helper function to register the require on self.
+ *
+ * The require property is used by VSCode/code-server
+ * to load files.
+ *
+ * We extracted the logic into a function so that
+ * it's easier to test.
+ **/
+export function registerRequireOnSelf({ self, window, nlsConfig, options }: RegisterRequireOnSelfType) {
+  const errorMsgPrefix = "[vscode]"
+
+  if (!self) {
+    throw new Error(`${errorMsgPrefix} Could not register require on self. self is undefined.`)
+  }
+
+  if (!window) {
+    throw new Error(`${errorMsgPrefix} Could not register require on self. window is undefined.`)
+  }
+
+  if (!options || !options.csStaticBase) {
+    throw new Error(
+      `${errorMsgPrefix} Could not register require on self. options or options.csStaticBase is undefined or missing.`,
+    )
+  }
+
+  if (!nlsConfig) {
+    throw new Error(`${errorMsgPrefix} Could not register require on self. nlsConfig is undefined.`)
+  }
+
+  const requireOnSelf: RequireOnSelfType = {
+    // Without the full URL VS Code will try to load file://.
+    baseUrl: `${window.location.origin}${options.csStaticBase}/lib/vscode/out`,
+    recordStats: true,
+    paths: {
+      "vscode-textmate": `../node_modules/vscode-textmate/release/main`,
+      "vscode-oniguruma": `../node_modules/vscode-oniguruma/release/main`,
+      xterm: `../node_modules/xterm/lib/xterm.js`,
+      "xterm-addon-search": `../node_modules/xterm-addon-search/lib/xterm-addon-search.js`,
+      "xterm-addon-unicode11": `../node_modules/xterm-addon-unicode11/lib/xterm-addon-unicode11.js`,
+      "xterm-addon-webgl": `../node_modules/xterm-addon-webgl/lib/xterm-addon-webgl.js`,
+      "tas-client-umd": `../node_modules/tas-client-umd/lib/tas-client-umd.js`,
+      "iconv-lite-umd": `../node_modules/iconv-lite-umd/lib/iconv-lite-umd.js`,
+      jschardet: `../node_modules/jschardet/dist/jschardet.min.js`,
+    },
+    "vs/nls": nlsConfig,
+  }
+
+  // TODO@jsjoeio
+  // I'm not sure how to properly type cast this
+  // This might be our best bet
+  // Source: https://stackoverflow.com/a/30740935
+  type FixMeLater = any
+  ;(self.require as FixMeLater) = requireOnSelf
```

Comment:
We could remove this and return the object to make the function free of side effects. Then we'd just have something like:

```
self.require = getRequire()
```

Although this function is basically just wrapping a simple object right? 

What do you think about doing `require("vscode")` and testing it that way? Or wrap the whole try/catch in a function called `configureLoader` or something like that?

Reply:
> We could remove this and return the object to make the function free of side effects. 

Even better, I like this!

> What do you think about doing require("vscode") and testing it that way?

Eh...I wouldn't say I'm in favor. I mean with this way, we're able to isolate this logic and test as a unit. Plus doing it this way, we have some better error handling so we know what went wrong if this fails. Writing a separate test which does that could be done later though! I would vote it's out of the scope of this. Thoughts?

> wrap the whole try/catch in a function

I thought about doing this — calling this function in a separate try/catch block, but we need `nlsConfig` which would be block-scoped to the other try/catch block, unless I move the declaration to a `let` and move it outside the try/catch block. 

> function called configureLoader

Are you suggesting renaming the current function to `configureLoader`? If so, I'm onboard!

Reply:
Yeah that's fair, I keep being hesitant about adding new API only for use in tests because it increases the area for future bugs (for example you can now call `registerRequireOnSelf` with the wrong arguments or in files where it doesn't make sense which you couldn't do before).

Maybe for all the browser files that run themselves we should adopt a pattern like this:
```javascript
function someFunc(document) {
}
function someOtherFunc(window) {
}

try {
  someFunc(document)
  someOtherFunc(window)
} catch (error) {
  console.error(error.message)
}
```

That way we minimize the amount of code that uses globals and needs to be tested via the `require` method. So in this case it would be something like:

```
try {
  const nlsConfig = getNLSConfiguration(document)
  const options = getOptions()
  configureLoader(self, window.location.origin, nlsConfig, options)
} catch (error) {
  console.error(error)
}
```

Reply:
BTW in case it helps I'm calling it things like `loaderConfig` because setting this object on `self.require` is what configures the loader and tells it how to load files that get imported.

Reply:
> Maybe for all the top-level browser files that run themselves we should adopt a pattern like this

I like the idea of this. That would make it cleaner, easier to maintain & test. I'll refactor to use this approach. 

> BTW in case it helps I'm calling it things like loaderConfig because setting this object on self.require is what configures the loader and tells it how to load files that get imported

Yes! Super helpful. Thank you for explaining that (I figured it was _like_ that but wasn't certain).